File: mellowplayer/gui/dlg_select_service.py
from mellowplayer.qt import QtCore, QtGui, QtWidgets
from .forms import dlg_select_service_ui


HTML = '''
<h1>%(sv_name)s</h1>
<h2>Description</h2>
%(sv_desc)s
<h2>Service information</h2>
<p><strong>Maintainer: </strong><a href=%(maintainer_url)s>%(maintainer_name)s
</a></p>
<p><strong>Version: </strong>%(sv_version)s</p>
'''


class DlgSelectService(QtWidgets.QDialog):
    def __init__(self, parent):
        super().__init__(parent)
        self.ui = dlg_select_service_ui.Ui_Dialog()
        self.ui.setupUi(self)
        self.ui.buttonBox.button(self.ui.buttonBox.Ok).setText('Use service')
        self.setWindowTitle('Select service - Mellow Player')
        services = parent.services.plugins.values()
        for sv in sorted(services, key=lambda x: x.metadata.name):
            meta = sv.metadata
            item = QtWidgets.QListWidgetItem(meta.name, self.ui.listWidget)
            if meta.icon:
                item.setIcon(QtGui.QIcon(meta.icon))
            item.service = sv
            self.ui.listWidget.addItem(item)
        self.ui.listWidget.setCurrentRow(0)
        self.ui.pushButton.hide()  # not use at the moment

    @QtCore.pyqtSlot(QtWidgets.QListWidgetItem, QtWidgets.QListWidgetItem)
    def on_listWidget_currentItemChanged(self, item, _):
        sv = item.service
        desc = HTML % {
            'sv_name': sv.metadata.name,
            'sv_desc': sv.description,
            'maintainer_url': sv.metadata.maintainer_link,
            'maintainer_name': sv.metadata.maintainer_name,
            'sv_version': sv.metadata.version_str
        }
        self.ui.textBrowser.setText(desc)

    @classmethod
    def select_service(cls, parent):
        dlg = cls(parent)
        if dlg.exec_() == dlg.Accepted:
            return dlg.ui.listWidget.currentItem().service.metadata.name
        return None

File: app/dlgselectservice.cpp
//---------------------------------------------------------
//
// This file is part of MellowPlayer.
//
// MellowPlayer is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// MellowPlayer is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with MellowPlayer.  If not, see <http://www.gnu.org/licenses/>.
//
//---------------------------------------------------------

#include <QDialogButtonBox>
#include <QPushButton>
#include <QMessageBox>
#include <mellowplayer.h>
#include "dlgselectservice.h"
#include "ui_dlg_select_service.h"

bool cmp(const IStreamingService* a,
                    const IStreamingService* b)
{
    return a->metaData().name < b->metaData().name;
}


//---------------------------------------------------------
DlgSelectServices::DlgSelectServices(QWidget *parent):
    QDialog(parent),
    m_ui(new Ui::DialogSelectServices())
{
    m_ui->setupUi(this);

    int row = 0;
    QString activeService = QSettings().value("service", "").toString();
    int i = 0;
    StreamingServicesList plugins = Services::streamingServices()->plugins();
    qSort(plugins.begin(), plugins.end(), cmp);
    foreach(IStreamingService* plugin, plugins){
        QListWidgetItem* item = new QListWidgetItem(m_ui->listWidget);
        item->setText(plugin->metaData().name);
        item->setIcon(plugin->metaData().icon);
        m_ui->listWidget->addItem(item);

        if(plugin->metaData().name == activeService)
            row = i;

        ++i;
    }

    connect(m_ui->listWidget, SIGNAL(currentRowChanged(int)),
            this, SLOT(onCurrentRowChanged(int)));
    m_ui->listWidget->setCurrentRow(row);
}

//---------------------------------------------------------
DlgSelectServices::~DlgSelectServices()
{
    delete m_ui;
}

//---------------------------------------------------------
void DlgSelectServices::onCurrentRowChanged(int row)
{
    QListWidgetItem* item = m_ui->listWidget->item(row);
    IStreamingService* plugin = Services::streamingServices()->plugin(item->text());
    if(plugin)
    {
        m_ui->textBrowser->setHtml(
              QObject::tr("<h1>%1</h1>"
                          "<h2>Description</h2>"
                          "%2"
                          "<h2>Service information</h2>"
                          "<p><strong>Maintainer: </strong><a href=%3>%4</a></p>"
                          "<p><strong>Version: </strong>%5</p>")
              .arg(plugin->metaData().name)
              .arg(plugin->metaData().description)
              .arg(plugin->metaData().author_website)
              .arg(plugin->metaData().author)
              .arg(plugin->metaData().version));
        m_ui->labelFlash->setVisible(
            plugin->flashRequired() && !Services::hasFlash());
    }
}

//---------------------------------------------------------
QString DlgSelectServices::selectService(QWidget *parent)
{
    DlgSelectServices dlg(parent);
    if(dlg.exec() == QDialog::Accepted)
    {
        QString serviceName = dlg.m_ui->listWidget->currentItem()->text();
        return serviceName;
    }
    return "";
}
